### PR TITLE
saul_adc: Add conversion functionality

### PR DESCRIPTION
**Pull request by gebart**
### Contribution description

Adds an option to specify an additional conversion function to the adc_params.h saul_adc configuration struct. The conversion function should take a raw ADC sample and convert it to some other measurement.
Included in the PR is an example generic filter function to convert ADC reading to real voltage. Only the Mulle platform is configured yet, but it is quite straightforward to implement the configuration.

### Issues/PRs references

Depends on #8566, #8577, #8592, #8593

### CODE REVIEW

**Comment by gebart** [comment]
One caveat is that the sensor type is still determined by the driver (saul_adc), so the source is still listed as SENSE_ANALOG when enumerated, regardless whether it puts out values in raw ADC numbers or e.g. a temperature reading

File: sys/phydat/phydat.c [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2018 Eistec AB
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_sensif
```
Comment (aabadie) [inline sys/phydat/phydat.c]: weird, this group is not defined (git grep only return an `@ingroup driver_sensif` in phydat_str.c. Looks like there a typo somewhere or something missing.

File: boards/mulle/include/board.h [hunk 2]
```diff
@@ -26,6 +26,15 @@
 #include "mulle-nvram.h"
 #include "mtd.h"
 
+/**
+ * @name ADC sample conversion configuration
+ * @{
+ */
+#define BOARD_ADC_VREFH_DEFAULT          33000
```
Comment (aabadie) [inline boards/mulle/include/board.h]: maybe put parenthesis around values

File: tests/unittests/tests-phydat/tests-phydat.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#include "tests-phydat.h"
+
+#include "phydat.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static void test_phydat_fit(void)
+{
+    int val0 = 100000;
```
Comment (aabadie) [inline tests/unittests/tests-phydat/tests-phydat.c]: Might be interesting to put a comment explaining what this test is supposed to do.

File: drivers/include/periph/adc_conv.h [hunk 4]
```diff
@@ -69,6 +77,15 @@ void adc_conv_set_vrefh(int new_vrefh);
...
 void adc_conv_set_vrefl(int new_vrefl);
 
+/**
+ * @brief Set the ADC VREF unit scale exponent
+ *
+ * This is the 10 exponent for the value in vrefh/vrefl, i.e.
+ * Real Vrefh = vrefh * 10^(scale)
+ * For example, scale = -3 means that vrefh, vrefl are given in millivolts.
+ */
+void adc_conv_set_scale(int new_scale);
```
[the remark this answers is not in the document]
Comment (gebart) [reply drivers/include/periph/adc_conv.h]: addressed in #8593

File: tests/unittests/tests-phydat/tests-phydat.c (hunk 3 above)
Comment (gebart) [reply tests/unittests/tests-phydat/tests-phydat.c]: addressed in #8592

File: boards/mulle/include/board.h (hunk 2 above)
Comment (gebart) [reply boards/mulle/include/board.h]: addressed

File: sys/phydat/phydat.c (hunk 1 above)
Comment (gebart) [reply sys/phydat/phydat.c]: addressed in #8592

**Comment by vincent-d** [comment]
This feature is interesting and I'm wondering whether it could be extended to work with any type of sensor within saul. My use case would be to convert a raw adc value form an i2c ADC to a voltage.

**Comment by gebart** [comment]
@vincent-d I agree that it might be useful, but might be better then to create a saul filter device which wraps another saul device.
Anyway, I think for these more advanced scenarios it will be very helpful to have the XFA feature (arrays distributed across different compilation units) #7523.

**Pull request closed without merging**